app.py

```diff
@@ -0,0 +1,162 @@
+from cryptacular.bcrypt import BCRYPTPasswordManager
+
+import sqlalchemy as sa
+from sqlalchemy.orm import scoped_session, sessionmaker
+from zope.sqlalchemy import ZopeTransactionExtension
+from sqlalchemy.ext.declarative import declarative_base
+
+from waitress import serve
+
+HERE = os.path.dirname(os.path.abspath(__file__))
+DBSession = scoped_session(sessionmaker(extension=ZopeTransactionExtension()))
+DATABASE_URL = os.environ.get(
+    'DATABASE_URL',
+    'postgresql://power_user:hownowbrownsnake@localhost:5432/test1'
+    #'postgresql://power_user:nopassword@localhost:5432/test1'
+)
+Base = declarative_base()
+
+
+class User(Base):
+    __tablename__ = 'users'
+
+    id = sa.Column(sa.Integer, primary_key=True, autoincrement=True)
+    username = sa.Column(sa.Unicode(127), nullable=False, unique=True)
+    password = sa.Column(sa.Unicode(127), nullable=False)
+
+    @classmethod
+    def new(cls, username=None, password=None, session=DBSession):
+        manager = BCRYPTPasswordManager()
+        if not (username and password):
+            raise ValueError("Username and password needed")
+        hashed = manager.encode(password)
+        instance = cls(username=username, password=hashed)
+        session.add(instance)
+        return instance
+
+    @classmethod
+    def get_by_id(cls, id, session=DBSession):
+        return session.query(cls).filter(cls.id == id).one()
+
+    @classmethod
+    def get_by_username(cls, username, session=DBSession):
+        return session.query(cls).filter(cls.username == username).one()
+
+
+@view_config(route_name="login", renderer="templates/gatepage.jinja2")
+def login_page(request):
+    if request.method == "POST":
+        username = request.params.get('username', '')
+        authenticated = False
+        try:
+            authenticated = login(request)
+        except ValueError as e:
+            return {'error': e}
+
+        if authenticated:
+            headers = remember(request, username)
+            return HTTPFound(request.route_url('home'), headers=headers)
+    else:
+        return {}
+
+
+@view_config(route_name="new_account", renderer="templates/new_account.jinja2")
+def new_account_page(request):
+    error = ""
+    if request.method == "POST":
+        try:
+            username = request.params.get('username', None)
+            passsword = request.params.get('password', None)
+            User.new(username, passsword)
+        except Exception as e:
+            error = e
+        return HTTPFound(request.route_url('home'))
+    return {'error': error}
+
+
+def login(request):
+    username = request.params.get("username", None)
+    password = request.params.get("password", None)
+    if not (username and password):
+        raise ValueError("Username and password are required")
+    manager = BCRYPTPasswordManager()
+    try:
+        user = User.get_by_username(username)
+    except:
+        raise ValueError("User does not exist")
```

Comment:
Sorry, sloppy.  Instead I should have said "You should learn what exception might be raised by not finding the user so that you don't use a bare `except` here".  Bare excepts are anathema.